src/mqtt/transport/tcp/mqtt_tcp.c

```diff
@@ -161,7 +161,7 @@ mqtt_pipe_timer_cb(void *arg)
 	// send PINGREQ with tmaio itself?
 	// nng_msleep(p->keepalive);
 	nni_mtx_lock(&p->mtx);
-	if (!p->busy) {
+	if (!p->busy && !nni_aio_busy(p->qsaio)) {
 		// send pingreq
 		buf[0] = 0xC0;
 		buf[1] = 0x00;
```

Comment:
Did you forget to set the pipe busy?
Should we also check if the aio is busy or not executed?